crates/red_knot_python_semantic/src/semantic_index.rs

```diff
@@ -240,6 +240,45 @@ impl<'db> SemanticIndex<'db> {
         Some(&self.scopes[self.parent_scope_id(scope_id)?])
     }
 
+    fn is_scope_reachable(&self, db: &'db dyn Db, scope_id: FileScopeId) -> bool {
+        match self.parent_scope_id(scope_id) {
+            None => true,
+            Some(parent_scope_id) => {
+                if !self.is_scope_reachable(db, parent_scope_id) {
+                    return false;
+                }
+
+                let parent_use_def = self.use_def_map(parent_scope_id);
+                let reachability = self.scope(scope_id).reachability();
+
+                parent_use_def.is_reachable(db, reachability)
+            }
+        }
```

Comment:
nit

```suggestion
        self.parent_scope_id(scope_id)
            .is_none_or(|parent_scope_id| {
                if !self.is_scope_reachable(db, parent_scope_id) {
                    return false;
                }

                let parent_use_def = self.use_def_map(parent_scope_id);
                let reachability = self.scope(scope_id).reachability();

                parent_use_def.is_reachable(db, reachability)
            })
```

Reply:
it might also be worth adding a comment saying something along the lines of "if `self.parent_scope_id()` is `None`, it's the global scope, which is always reachable"